sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/mongodb/MongoDbReadWriteIT.java

```diff
@@ -187,7 +199,68 @@ public void testWriteAndRead() {
     PCollection<Row> output =
         BeamSqlRelUtils.toPCollection(readPipeline, sqlEnv.parseQuery("select * from TEST"));
 
-    assertEquals(output.getSchema(), SOURCE_SCHEMA);
+    assertEquals(SOURCE_SCHEMA, output.getSchema());
+
+    PAssert.that(output).containsInAnyOrder(testRow);
+
+    readPipeline.run().waitUntilFinish();
+  }
+
+  @Test
+  public void testProjectPushDown() {
+    final Schema expectedSchema =
+        Schema.builder()
+            .addNullableField("c_varchar", STRING)
+            .addNullableField("c_boolean", BOOLEAN)
+            .addNullableField("c_integer", INT32)
+            .build();
+    Row testRow = row(expectedSchema, "varchar", true, 2147483647);
+
+    String createTableStatement =
+        "CREATE EXTERNAL TABLE TEST( \n"
+            + "   c_bigint BIGINT, \n "
+            + "   c_tinyint TINYINT, \n"
+            + "   c_smallint SMALLINT, \n"
+            + "   c_integer INTEGER, \n"
+            + "   c_float FLOAT, \n"
+            + "   c_double DOUBLE, \n"
+            + "   c_boolean BOOLEAN, \n"
+            + "   c_varchar VARCHAR, \n "
+            + "   c_arr ARRAY<VARCHAR> \n"
+            + ") \n"
+            + "TYPE 'mongodb' \n"
+            + "LOCATION '"
+            + mongoSqlUrl
+            + "'";
+    sqlEnv.executeDdl(createTableStatement);
+
+    String insertStatement =
+        "INSERT INTO TEST VALUES ("
+            + "9223372036854775807, "
+            + "127, "
+            + "32767, "
+            + "2147483647, "
+            + "1.0, "
+            + "1.0, "
+            + "TRUE, "
+            + "'varchar', "
+            + "ARRAY['123', '456']"
+            + ")";
+
+    BeamRelNode insertRelNode = sqlEnv.parseQuery(insertStatement);
+    BeamSqlRelUtils.toPCollection(writePipeline, insertRelNode);
+    writePipeline.run().waitUntilFinish();
+
+    BeamRelNode node = sqlEnv.parseQuery("select c_varchar, c_boolean, c_integer from TEST");
+    // Calc should be dropped, since MongoDb supports project push-down and field reordering.
+    assertThat(node, instanceOf(BeamPushDownIOSourceRel.class));
```

Comment:
I'm not crazy about this method of checking that the project push down happened, the fact that we produce a `BeamPushDownIOSourceRel` should be an implementation detail as far as this integration test is concerned.

Ideally we would get access to the list of fields retrieved from MongoDB somehow and make assertions about that instead. Is there any way to get that information? I don't know if this embedded mongo server supports that, but maybe some alternative (fongo?) does?

This isn't that big of a deal, if it's going to be a huge pain to re-work I'm fine leaving it as-is, but I think it's worth looking into.

Reply:
I see the benefit of checking what actually is being retrieved.

Fields retrieved from MongoDB are returned as a `PCollection<Document>` from `MongoDBIO.expand()`, which is called by `MongoDbTable.buildIOReader(...)`.
An `IOSourceRel` does not do any data manipulations (just wraps an IO and does type conversion).
I am not exactly sure that checking values returned by `MongoDBIO.expand()` would be much different, unless I am misunderstanding something.

I also [looked](https://stackoverflow.com/questions/41550450/mongodb-list-previous-queries) into retrieving some sort of log (from the embedded database) to check what query actually got executed, but I am not sure that we can easily retrieve that data.

Alternatively, we could validate that `withQueryFn` on `MongoDbIO.Read` does indeed have projected fields set.

What do you think?

Reply:
> I also looked into retrieving some sort of log (from the embedded database) to check what query actually got executed, but I am not sure that we can easily retrieve that data.

This is the sort of thing I had in mind. Making an assertion about the query fn is a little better, but I still think that should be an implementation detail from the perspective of this test.

We're in sort of a weird position, something like a mock mongodb client would make this really easy, but that won't work for us since we need to be able to reference a hostname and port for the mongo instance. 

Your link about setting the profile level seems promising - it looks like if we do that we could get a list of operations/queries from the [`system.profile` collection](https://docs.mongodb.com/manual/tutorial/manage-the-database-profiler/#view-profiler-data). We could query that and assert there's just one operation with the expected projection. That seems like the best bet to me. It's kind of a pain though, so if you want to punt on it I'd be fine with just filing a jira describing this approach and referencing it here.

Reply:
Makes sense. Ended up validating projections logged in `system.profile` collection. Thanks for your help!